### PR TITLE
Feat/migrate sql gen to macro

### LINKED ISSUE #457: Add missing features such as merge_exclude_columns and incremental_predicates
### Describe the feature
The current implementation of the adapter generates PySpark code with the corresponding queries in the impl.py file. I believe it would be better to handle the compiled SQL generation at the macro level. Generating PySpark code in impl.py has the following drawbacks:

- We do not utilize dbt's built-in compilation layer.
- The impl.py file becomes difficult to read and debug.
- The code found in the dbt/target/ folder does not match what was actually executed in the Spark session.

Additionally, we force users to use the glue_catalog namespace to query Iceberg tables. However, there is a solution to avoid this.

Finally, the following features are missing in the current catalog but can be easily implemented if we move SQL generation to the macro level:

- incremental_predicates
- merge_exclude_columns

### Describe alternatives you've considered
By handling the SQL generation at the macro level, we can benefit from the following:
- The executed code will match what was actually run in the Spark session.
- SQL templating and generation can be fully managed at the macro level, making the impl.py file more readable and maintainable.

We can use the SparkSessionCatalog as an implementation of the Spark catalog. More information can be found in the following link [Iceberg Catalog configuration](https://iceberg.apache.org/docs/latest/spark-configuration/#catalog-configuration). I tested the following configruation and it worked well :
```
--conf spark.sql.catalog.spark_catalog=org.apache.iceberg.spark.SparkSessionCatalog
--conf spark.sql.catalog.spark_catalog.warehouse=s3://al-gdo-dev-ww-dl-0139-transfo/data
--conf spark.sql.catalog.spark_catalog.catalog-impl=org.apache.iceberg.aws.glue.GlueCatalog
--conf spark.sql.catalog.spark_catalog.io-impl=org.apache.iceberg.aws.s3.S3FileIO
--conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions
```
_One thing to note about this configuration, **it makes CTAS and RTAS operations non atomic**_

### Are you interested in contributing this feature?
I have already implemented this on my side. I can create a PR and let you check it @moomindani if it's okey for you.

### PR DESCRIPTION
**Pull request by aiss93**
resolves #457

### Description

The following PR addresses the following points:

- Migrates the PySpark code for the Iceberg file format at a macro level, making the impl.py file more readable.
- Fixes the get_columns_in_relation function to work for both Iceberg and non-Iceberg tables without hard-coding the catalog name.
- Fixes the get_location table function to work for both Iceberg and non-Iceberg tables on macOS and Windows. The old os.path.join function didn’t work well on Windows OS, as it built a path using '' instead of '/', leading to S3 path location errors.
- Adds a helper function to retrieve the Iceberg catalog namespace from the profile.yaml file, allowing users to name their Iceberg catalog namespace as desired.
- Adds merge_exclude_columns and incremental_predicates features.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-glue next" section.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by moomindani** [comment]
@aiss93 Apologizes I forgot to submit my comments last week. Submitted them now. Having test_iceberg is good idea, but I think we still need to make sure that we have enough coverage in the test cases. Plz refer to my comments.

File: dbt/include/glue/macros/adapters.sql [hunk 1]
```diff
@@ -59,50 +67,37 @@
...
+    {{ glue__location_clause() }}
+    {{ comment_clause() }}
     as
-      {{ sql }}
-  {% endcall %}
+    {{ sql }}
+  {%- endif %}
 {%- endmacro -%}
 
 {% macro glue__snapshot_get_time() -%}
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/adapters.sql:101]: This path existed before the PR and wan't tested.

File: dbt/include/glue/macros/adapters.sql [hunk 2]
```diff
@@ -59,50 +67,37 @@
 {%- endmacro -%}
 
 {% macro glue__create_table_as(temporary, relation, sql) -%}
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/adapters.sql:67]: This path existed before the PR and wan't tested.

File: dbt/adapters/glue/impl.py [hunk 3]
```diff
@@ -885,167 +853,22 @@ def hudi_merge_table(self, target_relation, request, primary_key, partition_key,
...
-        try:
-            session.cursor().execute(code)
-        except DbtDatabaseError as e:
-            raise DbtDatabaseError(msg="GlueIcebergWriteTableFailed") from e
-        except Exception as e:
-            logger.error(e)
-
     @available
-    def iceberg_expire_snapshots(self, table):
+    def iceberg_expire_snapshots(self, relation):
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/adapters/glue/impl.py:874]: This function existed before the PR and wan't tested.

File: dbt/include/glue/macros/materializations/incremental/incremental.sql [hunk 4]
```diff
@@ -1,100 +1,89 @@
 {% materialization incremental, adapter='glue' -%}
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/materializations/incremental/incremental.sql:1]: This path is actually tested in the iceberg functional tests. Precisely in the TestIncrementalGlue.

File: dbt/include/glue/macros/materializations/incremental/validate.sql [hunk 5]
```diff
@@ -0,0 +1,50 @@
+{% macro dbt_glue_validate_get_file_format(raw_file_format) %}
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/materializations/incremental/validate.sql:1]: This function comes from the Spark SQL adapter. To remove the dependecy between the two adapter I added it as an utility macro. This macro was used before but not tested as well.

File: dbt/include/glue/macros/materializations/incremental/incremental.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (moomindani) [reply dbt/include/glue/macros/materializations/incremental/incremental.sql:1]: Sounds good

File: tests/unit/test_credentials.py [hunk 6]
```diff
@@ -16,3 +16,4 @@ def test_credentials(self) -> None:
         assert credentials.schema == "tests"
         assert credentials.database is None
         assert credentials.glue_version == "4.0"    # default Glue version is 4.0
+        assert credentials.custom_iceberg_catalog_namespace == "glue_catalog"
```
[the remark this answers is not in the document]
Comment (aiss93) [reply tests/unit/test_credentials.py:19]: Will add

File: dbt/adapters/glue/impl.py [hunk 7]
```diff
@@ -885,167 +853,22 @@ def hudi_merge_table(self, target_relation, request, primary_key, partition_key,
         except Exception as e:
             logger.error(e)
 
-    def iceberg_create_or_replace_table(self, target_relation, partition_by, table_properties):
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/adapters/glue/impl.py:888]: Already included in the iceberg functionnal tests.

[the remark this answers is not in the document]
Comment (moomindani) [reply dbt/adapters/glue/impl.py:888]: Looks great.

File: dbt/include/glue/macros/materializations/incremental/strategies.sql [hunk 8]
```diff
@@ -16,16 +16,60 @@
...
+            {{ column_name }} = DBT_INTERNAL_SOURCE.{{ column_name }}
+            {%- if not loop.last %}, {%- endif %}
+        {%- endfor %}
+        {%- else %} * {% endif %}
+
+      when not matched then insert *
+{% endmacro %}
+
+
+{% macro dbt_glue_get_incremental_sql(strategy, source, target, unique_key, incremental_predicates) %}
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/materializations/incremental/strategies.sql:63]: Already included in the iceberg functionnal tests.

File: dbt/include/glue/macros/materializations/incremental/strategies.sql [hunk 9]
```diff
@@ -16,16 +16,60 @@
...
   {%- if strategy == 'append' -%}
     {#-- insert new records into existing table, without updating or overwriting #}
     {{ get_insert_into_sql(source, target) }}
   {%- elif strategy == 'insert_overwrite' -%}
     {#-- insert statements don't like CTEs, so support them via a temp view #}
     {{ get_insert_overwrite_sql(source, target) }}
   {%- elif strategy == 'merge' -%}
-  {#-- merge all columns with databricks delta - schema changes are handled for us #}
-    {{ get_merge_sql(target, source, unique_key, dest_columns=none, predicates=none) }}
+    {#-- merge all columns with databricks delta - schema changes are handled for us #}
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/materializations/incremental/strategies.sql:71]: will change comments

File: dbt/include/glue/macros/adapters.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/adapters.sql:67]: Apologies. This path is actually testes i both IcebergTests and BasicTests. This path is called in most tests

[the remark this answers is not in the document]
Comment (moomindani) [reply dbt/include/glue/macros/adapters.sql:67]: Then that is fine, thanks for verifying it.

File: dbt/include/glue/macros/materializations/incremental/strategies.sql [hunk 10]
```diff
@@ -16,16 +16,60 @@
     select {{dest_cols_csv}} from {{ source_relation }}
 {% endmacro %}
 
-{% macro dbt_glue_get_incremental_sql(strategy, source, target, unique_key) %}
+
+{% macro get_merge_sql(target, source, unique_key, dest_columns, incremental_predicates) %}
```
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/materializations/incremental/strategies.sql:20]: This path is tested in the IcebergIncremntalTest

[the remark this answers is not in the document]
Comment (moomindani) [reply dbt/include/glue/macros/materializations/incremental/strategies.sql:20]: nice

File: dbt/include/glue/macros/adapters.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (moomindani) [reply dbt/include/glue/macros/adapters.sql:101]: nice.

File: tests/unit/test_credentials.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aiss93) [reply tests/unit/test_credentials.py:19]: Done. Please check the note section of the iceberg part in the readme

File: dbt/include/glue/macros/materializations/incremental/strategies.sql (hunk 9 above)
[the remark this answers is not in the document]
Comment (aiss93) [reply dbt/include/glue/macros/materializations/incremental/strategies.sql]: Comment removed

**Comment by aiss93** [comment]
Hi @moomindani do you have any additional comments or can we launch the functional tests ?
Thank you for your time

**Comment by moomindani** [comment]
> Hi @moomindani do you have any additional comments or can we launch the functional tests ? Thank you for your time

Thanks for making changes based on my comments. Apologizes for delay in review. We are still reviewing the changes. It may take some time as our bandwidth is very limited this month, but let us update you.

**Comment by aiss93** [comment]
Sure thank you !

**Review by moomindani — APPROVED** [review]

**Comment by moomindani** [comment]
Thank you for your contribution! This patch brought huge value, we really appreciate your contribution.

File: README.md [hunk 11]
```diff
@@ -607,20 +607,21 @@ dbt will run an [atomic `merge` statement](https://iceberg.apache.org/docs/lates
...
+
 ```
---conf spark.sql.catalog.AnotherRandomCatalogName=org.apache.iceberg.spark.SparkCatalog
+--conf spark.sql.catalog.{catalog_namespace}=org.apache.iceberg.spark.SparkCatalog
+--conf spark.sql.catalog.{catalog_namespace}.warehouse={warehouse_path}
+--conf spark.sql.catalog.{catalog_namespace}.catalog-impl=org.apache.iceberg.aws.glue.GlueCatalog
+--conf spark.sql.catalog.{catalog_namespace}.io-impl=org.apache.iceberg.aws.s3.S3FileIO
 ```
-In both cases, the underlying catalog would be the AWS Glue Catalog, unique in your AWS Account and Region, and you would be able to work with the exact same data. Also make sure that if you change the name of the Glue Catalog Alias, you change it in all the other `--conf` where it's used: 
+When using the default value, the following spark configruation should be added to enable iceberg.
```
Comment (sanga8) [inline README.md:619]: configuration

Comment (moomindani) [reply README.md:619]: Oops thanks I did not notice. Addressed it.

**Pull request merged**